File: apps/desktop-tauri/studio-ui/src/editor/gradeKernel/hsl.ts
// RGB ↔ HSL conversion (mirror Rust `ops/hsl.rs`).

import type { Rgb } from "./types";

/** RGB (`0..=1`) → HSL with hue in degrees (mirrors Rust `rgb_to_hsl`). */
export function rgbToHsl(rgb: Rgb): [number, number, number] {
  const max = Math.max(...rgb);
  const min = Math.min(...rgb);
  const l = (max + min) / 2;
  const d = max - min;
  if (d <= 0) return [0, 0, l];
  const s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
  let h: number;
  if (max === rgb[0]) h = 60 * ((((rgb[1] - rgb[2]) / d) % 6 + 6) % 6);
  else if (max === rgb[1]) h = 60 * ((rgb[2] - rgb[0]) / d + 2);
  else h = 60 * ((rgb[0] - rgb[1]) / d + 4);
  return [h, s, l];
}

/** HSL (hue in degrees) → RGB (mirrors Rust `hsl_to_rgb`). */
export function hslToRgb(h: number, s: number, l: number): Rgb {
  const c = (1 - Math.abs(2 * l - 1)) * s;
  const hp = h / 60;
  const x = c * (1 - Math.abs(((hp % 2) + 2) % 2 - 1));
  const seg = Math.min(Math.floor(hp), 5);
  const [r, g, b] = [
    [c, x, 0],
    [x, c, 0],
    [0, c, x],
    [0, x, c],
    [x, 0, c],
    [c, 0, x],
  ][seg];
  const m = l - c / 2;
  return [r + m, g + m, b + m];
}

File: crates/hgripe-grade/src/ops/hsl.rs
// RGB ↔ HSL conversion (hue in degrees) — the colour model behind the HSL
// ops, HSL curves, the colour warper and the qualifier.

/// RGB (`0..=1`) → HSL with hue in degrees (`0..360`).
pub(crate) fn rgb_to_hsl(rgb: [f32; 3]) -> (f32, f32, f32) {
    let max = rgb[0].max(rgb[1]).max(rgb[2]);
    let min = rgb[0].min(rgb[1]).min(rgb[2]);
    let l = (max + min) / 2.0;
    let d = max - min;
    if d <= 0.0 {
        return (0.0, 0.0, l);
    }
    let s = if l > 0.5 {
        d / (2.0 - max - min)
    } else {
        d / (max + min)
    };
    let h = if max == rgb[0] {
        60.0 * ((rgb[1] - rgb[2]) / d).rem_euclid(6.0)
    } else if max == rgb[1] {
        60.0 * ((rgb[2] - rgb[0]) / d + 2.0)
    } else {
        60.0 * ((rgb[0] - rgb[1]) / d + 4.0)
    };
    (h, s, l)
}

/// HSL (hue in degrees) → RGB (`0..=1`).
pub(super) fn hsl_to_rgb(h: f32, s: f32, l: f32) -> [f32; 3] {
    let c = (1.0 - (2.0 * l - 1.0).abs()) * s;
    let hp = h / 60.0;
    let x = c * (1.0 - (hp.rem_euclid(2.0) - 1.0).abs());
    let (r, g, b) = match hp as u32 {
        0 => (c, x, 0.0),
        1 => (x, c, 0.0),
        2 => (0.0, c, x),
        3 => (0.0, x, c),
        4 => (x, 0.0, c),
        _ => (c, 0.0, x),
    };
    let m = l - c / 2.0;
    [r + m, g + m, b + m]
}
